unit/unit_reader/unit_reader.c

```diff
@@ -1040,6 +1042,69 @@ test_vec(
     gbinder_ipc_unref(ipc);
 }
 
+/*==========================================================================*
+ * byte_array
+ *==========================================================================*/
+
+static
+void
+test_byte_array(
+    void)
+{
+    const char in_data[] = "1234abcd";
+    gint32 in_len = sizeof(in_data) - 1;
+
+    const void *out_data = NULL;
+    gint32 out_len = 0;
+
+    void *tmp;
+    gsize tmp_len = sizeof(in_len) + in_len;
+
+    gint32 null_len = -1;
+
+    GBinderDriver* driver;
+    GBinderReader reader;
+    GBinderReaderData data;
+
+    /* test for NULL */
+    g_assert((driver = gbinder_driver_new(GBINDER_DEFAULT_BINDER, NULL)));
+
+    tmp = g_malloc0(sizeof(null_len));
+    memcpy(tmp, &null_len, sizeof(null_len));
+
+    memset(&data, 0, sizeof(data));
+    data.buffer = gbinder_buffer_new(driver, tmp, sizeof(null_len));
+
+    gbinder_reader_init(&reader, &data, 0, sizeof(null_len));
+
+    g_assert(!(out_data = gbinder_reader_read_byte_array(&reader, &out_len)));
+    g_assert(gbinder_reader_at_end(&reader));
+    g_assert(null_len == out_len);
+
+    gbinder_buffer_free(data.buffer);
+    gbinder_driver_unref(driver);
+
+    /* test for data */
+    g_assert((driver = gbinder_driver_new(GBINDER_DEFAULT_BINDER, NULL)));
+
+    tmp = g_malloc0(tmp_len);
+    memcpy(tmp, &in_len, sizeof(in_len));
+    memcpy(tmp + sizeof(in_len), in_data, in_len);
+
+    memset(&data, 0, sizeof(data));
+    data.buffer = gbinder_buffer_new(driver, tmp, tmp_len);
+
+    gbinder_reader_init(&reader, &data, 0, tmp_len);
+
+    g_assert((out_data = gbinder_reader_read_byte_array(&reader, &out_len)));
+    g_assert(gbinder_reader_at_end(&reader));
+    g_assert(in_len == out_len);
+    g_assert(memcmp(in_data, out_data, in_len) == 0);
+
+    gbinder_buffer_free(data.buffer);
+    gbinder_driver_unref(driver);
```

Comment:
One negative branch in `gbinder_reader_read_byte_array()` (when `gbinder_reader_can_read()` at line 572 returns FALSE) is not covered:
```
 572   [ +  - ]:    1 :   } else if (gbinder_reader_can_read(p, sizeof(*ptr) + *ptr)) {
```